File: 0-9999/2000-2999/2400-2499/2428/2428.py
from typing import List 

def solution(N:int, files:List[int]) -> int:
    files.sort()
    
    result = 0
    left = 0
    for right in range(N):
        # left는 전체 for문(즉, 전체 코드)에서 0부터 N-1까지 한 번씩만 증가
        # O(2N) -> 각각 한 방향으로만 움직임
        while left < right and files[left] < 0.9 * files[right]:
            left += 1
        result += (right - left)
    return result
    
if __name__ == '__main__':
    N = int(input())    
    
    files = list(map(int,input().split()))
    
    print(solution(N=N, files=files))

File: 0-9999/2000-2999/2400-2499/2428/2428.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func solution(N int, files []int) int {

	sort.Slice(files, func(i, j int) bool {
		return files[i] < files[j]
	})

	result := 0

	left := 0
	for right := 0; right < N; right++ {
		for left < right && float64(files[left]) < 0.9*float64(files[right]) {
			left += 1
		}
		// left == right 인 경우 제외
		result += (right - left)
	}
	return result
}

func main() {
	reader := bufio.NewReader(os.Stdin)

	line1, _ := reader.ReadString('\n')
	N, _ := strconv.Atoi(strings.TrimSpace(line1))

	line2, _ := reader.ReadString('\n')
	tokens := strings.Fields(line2)

	files := make([]int, N)
	for i := 0; i < N; i++ {
		files[i], _ = strconv.Atoi(tokens[i])
	}

	fmt.Println(solution(N, files))
}
